src/server/game/Combat/ThreatManager.h

```diff
@@ -15,284 +14,154 @@
-
-        // Reset all aggro without modifying the threadlist.
-        void resetAllAggro();
-
-        // Reset all aggro of unit in threadlist satisfying the predicate.
-        template<class PREDICATE> void resetAggro(PREDICATE predicate)
-        {
-            ThreatContainer::StorageType &threatList = iThreatContainer.iThreatList;
-            if (threatList.empty())
-                return;
-
-            for (ThreatContainer::StorageType::iterator itr = threatList.begin(); itr != threatList.end(); ++itr)
-            {
-                HostileReference* ref = (*itr);
-
-                if (predicate(ref->getTarget()))
-                {
-                    ref->setThreat(0);
-                    setDirty(true);
-                }
-            }
-        }
-
-        // methods to access the lists from the outside to do some dirty manipulation (scriping and such)
-        // I hope they are used as little as possible.
-        ThreatContainer::StorageType const & getThreatList() const { return iThreatContainer.getThreatList(); }
-        ThreatContainer::StorageType const & getOfflineThreatList() const { return iThreatOfflineContainer.getThreatList(); }
-        ThreatContainer& getOnlineContainer() { return iThreatContainer; }
-        ThreatContainer& getOfflineContainer() { return iThreatOfflineContainer; }
+        static const uint32 CLIENT_THREAT_UPDATE_INTERVAL = 1000u;
+
+        static const CompareThreatLessThan CompareThreat;
+        static bool CompareReferencesLT(ThreatReference const* a, ThreatReference const* b, float aWeight);
+
+        static float CalculateModifiedThreat(float threat, Unit const* victim, SpellInfo const* spell);
+
+        ThreatManager(Unit* owner);
+        void Update(uint32 tdiff); // called from Creature::Update (only creatures can have their own threat list)
+
+        Unit* GetOwner() const { return _owner; }
+        Unit* GetCurrentVictim() const { return _currentVictimRef->GetVictim(); }
+        Unit* GetAnyTarget() const;
+        Unit* SelectVictim();
+
+        bool IsEngaged() const { return _ownerEngaged; }
+        bool IsThreatListEmpty(bool includeOffline = false) const;
+        bool IsThreatenedBy(Unit const* who, bool includeOffline = false) const;
+        float GetThreat(Unit const* who, bool includeOffline = false) const;
+        std::vector<ThreatReference*> const& GetThreatList() const { return _sortedThreatList; } // NB: Not necessarily structured in any way unless you call HeapifyThreatList()/SortHeapifiedThreatList() first!
+
+        bool IsThreateningAnyone(bool includeOffline = false) const;
+        std::unordered_map<ObjectGuid, ThreatReference*> const& GetThreatenedByMeList() const { return _threatenedByMe; }
+
+        void UpdateOnlineStates(bool meThreateningOthers = true, bool othersThreateningMe = true);
+        ///== AFFECT MY THREAT LIST ==
+        void HeapifyThreatList();
+        void SortHeapifiedThreatList();
```

Comment:
Interesting, `boost::heap` provides mutability interfaces (which were lacking in `std::priority_queue`). Will look into it, could be a very nice improvement (fibonacci heaps give us constant complexity on increasing element priority, which is by far the most common - threat being added).